pkg/endpoint/endpoint.go

```diff
@@ -429,6 +573,10 @@ func NewEndpointFromChangeModel(base *models.EndpointChangeRequest, l pkgLabels.
 		}
 	}
 
+	ep.Mutex.Lock()
+	defer ep.Mutex.Unlock()
+	ep.runK8sCiliumEndpointSync()
```

Comment:
I understand the desire of this but I think we need to move it somewhere else, in code like this:

``` go
        addLabels := labels.ParseStringLabels(lbls)
        ep, err := endpoint.NewEndpointFromChangeModel(epTemplate, addLabels)
        if err != nil {
                return PutEndpointIDInvalidCode, err
        }
        ep.SetDefaultOpts(d.conf.Opts)

        oldEp, err2 := endpointmanager.Lookup(id)
        if err2 != nil {
                return PutEndpointIDInvalidCode, err2
        } else if oldEp != nil {
                return PutEndpointIDExistsCode, fmt.Errorf("Endpoint ID %s exists", id)
        }
        if err = endpoint.APICanModify(ep); err != nil {
                return PutEndpointIDInvalidCode, err
        }

        if err := endpointmanager.AddEndpoint(d, ep, "Create endpoint from API PUT"); err != nil {
                log.WithError(err).Warn("Aborting endpoint join")
                return PutEndpointIDFailedCode, err
        }
```

it will cause controllers to be leaked in the error paths because they never get `Leave()` invoked on them.

How about we do it in `endpointmanager.AddEndpoint()` instead? It guarantees that we call `Leave()` again. I don't think we need  controller for any endpoint that is not managed by the endpointmanager.

Reply:
This is much cleaner. I'll try to do it (I hope there isn't an import cycle).